kernel/src/main/java/org/kframework/parser/inner/disambiguation/CollapseRecordProdsVisitor.java

```diff
@@ -0,0 +1,82 @@
+ */
+public class CollapseRecordProdsVisitor extends SetsTransformerWithErrors<KEMException> {
+    @Override
+    public Either<Set<KEMException>, Term> apply(TermCons tc) {
+        if (tc.production().att().contains(Att.RECORD_PRD(), Production.class)) {
+            Production origPrd = tc.production().att().get(Att.RECORD_PRD(), Production.class);
+            Map<String, Term> children = new HashMap<>();
+            TermCons iterator = tc;
+
+            // find all named items
+            while (iterator != null && iterator.production().att().contains(Att.RECORD_PRD(), Production.class)) {
+                if (iterator.production().att().contains("recordPrd-main"))
+                    iterator = (TermCons) iterator.get(0);
+                else if (iterator.production().att().contains("recordPrd-one")) {
+                    String key = iterator.production().att().get("recordPrd-one");
+                    children.put(key, iterator.get(0));
+                    iterator = null;
+                } else if (iterator.production().att().contains("recordPrd-item")) {
+                    String key = iterator.production().att().get("recordPrd-item");
+                    if (children.containsKey(key))
+                        return Left.apply(Sets.newHashSet(KEMException.innerParserError("Duplicate record production key: " + key, tc)));
+                    else
+                        children.put(key, iterator.get(0));
+                    iterator = null;
+                } else if (iterator.production().att().contains("recordPrd-repeat")) {
+                    TermCons item = (TermCons) iterator.get(1);
+                    String key = item.production().att().get("recordPrd-item");
+                    if (children.containsKey(key))
+                        return Left.apply(Sets.newHashSet(KEMException.innerParserError("Duplicate record production key: " + key, tc)));
+                    else
+                        children.put(key, item.get(0));
+                    iterator = (TermCons) iterator.get(0);
+                } else
+                    iterator = null;
+            }
+            // construct expanded term with provided items and anonymous variables for missing ones
+            PStack<Term> collapsedItems = ConsPStack.empty();
+            for (NonTerminal nt : mutable(origPrd.nonterminals())) {
+                if (nt.name().isDefined() && children.containsKey(nt.name().get()))
+                    collapsedItems = collapsedItems.plus(children.get(nt.name().get()));
+                else {
+                    Production anonVarPrd = Production.apply(Seq(), Sorts.KVariable(), Seq(Terminal.apply("_[A-Z][A-Za-z0-9'_]*")), Att.empty());
+                    // The name is required so disambiguation doesn't collapse the variables into the same term.
+                    collapsedItems = collapsedItems.plus(Constant.apply("_" + nt.name().getOrElse(() -> "Gen") + uid++, anonVarPrd, tc.location(), tc.source()));
```

Comment:
This is not a sound way of generating names for variables, because the user could write a variable named _Gen0 in their rule. You need to either change it back to naming the variable "_" so the anonymous variables pass will give it a fresh name, or else you need to collect all the variables mentioned in the sentence and increment the counter inside a loop until you find one that isn't already used.